Java reflections

MightyCupcakes opened the pull request:
- [x] Overview
- [x] Basics
- [x] What reflections can do
- [x] Drawbacks/ Things to consider before using reflections

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+### What is Java reflections
+
+Java reflections allow the inspection of classes, methods and fields during runtime without having any knowledge of it
+during compile time. It is also possible to new instance of existing objects, invoke methods and changing values of
```
Can consider adding a comma after "during runtime". It also seems like a word is missing in the second sentence btw! It might be better to write "...change values..." instead of "...changing values...".

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+### The Basics of Reflections
+
+Before getting started with reflections in Java, it is important to realise that classes itself is an object. Every
+unique `Object` is assigned an immutable `Class` object by the JVM. This is fundamentally different from *instances* of
```
You may want to replace 'This' with 'This immutable `Class` object' as I think it may not be clear what 'This' is referring to. (Is is the fact that every object is assigned a class object? Is it the unique object? Or is it the class object?)

leeyimin commented on contents/java/Java.md:
````diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+}
+```
+
+An instance of a class can be created a usual using the new keyword
````
Typo here! May add colon to be consistent too?

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+There are many webpages dedicated to explaining the details of reflections in Java; so this will not repeat what is
+being made readily available on the web. One good place to start is [http://www.javaworld.com/article/2077015/java-se/take-an-in-depth-look-at-the-java-reflection-api.html](http://www.javaworld.com/article/2077015/java-se/take-an-in-depth-look-at-the-java-reflection-api.html)
+
+What this page will do however is to give some examples to demonstrate what reflections can do and in then give 
```
May consider adding comma before and after "however" or shift it to the front. There is a typo here too ("...and in then give you ideas...").

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+
+#### Applications
+
+You might have learnt from your Software Engineering module that the Observer pattern can be used for Objects that are
```
Hmm why is the Objects capitalized?

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+#### Applications
+
+You might have learnt from your Software Engineering module that the Observer pattern can be used for Objects that are
+interested to get notified if the state of another object is changed. However, a you don't want to create a troublesome
```
"However, a you don't want..." typo here!

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+bidirectional dependency between two unrelated objects that have no business talking to each other. This is when the
+Observer pattern becomes useful.
+
+One prime example of the implementation of the `Observer` pattern is the Google Events bus used in `AddressBook Level 4`.
```
It is probably better not to assume students have taken CS2103 specifically (The book is aimed at typical SE students, but not necessarily SE students in SoC. -- from the 3281 schedule page) and maybe just add a reference to the actual code with a link?

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+registered `Sheep extends Animal` as an event handler, both `Sheep` and `Animal` will be captured by the first line. The
+following lines will then examine all their methods (during runtime!) for the subscribe annotation and register it so
+that it will receive the specified event when it is fired. Of course this implementation leaves out a lot of details but
+you get the idea of how java reflections
```
The sentence ended prematurely?

leeyimin commented on contents/java/Java.md:
````diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+
+```java
+// TypeToken class is provided by Google Guava reflection library
+Set<? extends Class<?>> supertypes = TypeToken.of(clazz).getTypes().rawTypes();
````
Maybe you can instantiate a `Sheep` object here and give the object a more meaningful name here!

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+
+### Disadvantages of reflections
+
+While Java reflection is powerful, you should not immediately jump on the reflections ship. This is because there are
```
Maybe write "Java reflections" instead of "Java reflection" to be consistent with the plural form used throughout the chapter.

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+While Java reflection is powerful, you should not immediately jump on the reflections ship. This is because there are
+some drawbacks whenever you use reflections in your project.
+
+* Converts a compile-time error to a potentially destructive run-time error.
```
I find the use of point form quite abrupt here. Maybe add something like "Possible drawbacks include:"?

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+Compile time errors are easy to catch. Whenever you compile your code, Java's compiler cleverly spots any error you
+missed and points it out (along with line number and other useful information) to you before quitting. But by using
+reflections, you are bypassing these checks because there is no way to check such errors during compile time.
+Runtime errors are more troublesome in that they are harder to catch and debug and might even bring your whole software
```
Maybe give a short code example on a possible runtime error? I find this point a little abstract.

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+
+* Reflections are harder to understand and harder to debug
+
+There is a reason why reflections is placed under the advanced section. Codes using reflections are fundamentally
```
"...the topic of reflections is..." instead of "reflections is"? (seems like a subject-verb agreement mistake)

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+
+* Poor performance
+
+Since reflections resolves types dynamically, performance suffers. This is usually not an issue with small software
```
"Since reflections resolve" instead of "Since reflections resolves"? (same as above)

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+reflections? This should be very concerning if your software deals with sensitive information because 
+other classes can access fields that they are not supposed to.
+
+## Further Resources for reflections
```
Use ### instead of ##?

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+library](https://github.com/google/guava/wiki/ReflectionExplained) which contains many utility methods and classes to
+make your life easier.
+
+#### Applications
```
Maybe should use ### here!

leeyimin reviewed (changes requested):
An interesting read! Can possibly give one or two more applications (The heading is plural too haha).

MightyCupcakes committed:
Java Reflections: Add book chapter under advanced topics

MightyCupcakes committed:
Java Reflections: Add additional example

MightyCupcakes committed:
Java Reflections: Add more examples for clarity.

MightyCupcakes committed:
Java Reflections: Add author

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+
+### The Basics of Reflections
+
+Before getting started with reflections in Java, it is important to realise that classes itself is an object. Every
```
classes themselves are objects?

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+You might have learnt from your Software Engineering module that the observer pattern can be used for objects that are
+interested to get notified if the state of another object is changed. However, you don't want to create a troublesome
+bidirectional dependency between two unrelated objects that have no business talking to each other. This is when the
+observer pattern becomes useful.
```
You may consider restructuring this paragraph as the second sentence seems to imply there is something wrong with observer pattern and the reader would only understand the meaning of the sentence until he/she reads the third sentence. Maybe talk about the problem first, then the consideration of bidirectional dependency, then finally about observer pattern?

leeyimin commented on contents/java/Java.md:
````diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+public @interface Subscribe { }
+```
+
+And that is it! Important parts of the code to note is the first and second line before the declaration. The first line
````
Important parts ... to note are...?

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+        name = nameOfSheep;
+        EVENTS_BUS.register(this);
+    }
+    
```
maybe add a comment like ```// other methods here``` so that readers understand there are other methods in the class or add a method with the ```@Subscribe``` annotation? It is a little weird that you talk about this class later on and about the methods which are subscribed to but the full class here have no examples of methods like that.

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+class and converts it to a set. That is if you registered `Sheep extends Animal` as an event handler to the method,
+both `Sheep` and `Animal` will be captured by the first line. The following lines will then examine all their methods
+(during runtime!) for the `Subscribe` annotation and register it so that it will receive the specified event when it 
+is fired. Of course this implementation leaves out a lot of details but you get the idea of how java reflections works.
```
"...register it so that it will receive the specified event when it is fired." The "it"s are slightly ambiguous so maybe replace one of them with a more specific noun (```EventsCenter``` or "the method")?

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+(during runtime!) for the `Subscribe` annotation and register it so that it will receive the specified event when it 
+is fired. Of course this implementation leaves out a lot of details but you get the idea of how java reflections works.
+
+Another example of reflections might arise when you are trying to get a private field of another class. While this should
```
Can consider replacing "might arise when you are trying" with "is" or something else to be consistent with the usage of present tense.

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+optimally be solved by modifying the field visibility to `protected` or `public`, sometimes it is not possible to do so
+becuase you might not have any access to the code (for example codes in libraries or frameworks).
+
+For the sake of simplicity, let use the previous example of a `Animal`. The class definition can look like this.
```
let's instead of let

leeyimin commented on contents/java/Java.md:
````diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+}
+```
+
+And let us assume that you for some reason cannot modify this code. But you are interested in making a new class `Sheep`
````
"...you, for some reason, cannot..." may be better here

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+
+### Disadvantages of reflections
+
+While Java reflections is powerful, you should not immediately jump on the reflections ship. This is because there are
```
"While Java reflections are..." instead of "is"

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+some drawbacks whenever you use reflections in your project. The following are some points you should consider before
+using Java reflections:
+
+* Reflections converts a compile-time error to a potentially destructive run-time error.
```
convert instead of converts

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+
+Compile time errors are easy to catch. Whenever you compile your code, the compiler cleverly spots any error you
+missed and points it out (along with line number and other useful information) to you before quitting. But by using
+reflections, you are bypassing these checks because there is no way to check such errors during compile time. Runtime
```
"such errors" sounds like you are referring to compile time errors rather than errors in the code. Maybe add a sentence like "These uncaught errors may cause your program to fail during runtime instead, turning into runtime errors." so that the paragraph flows better.

leeyimin commented on contents/java/Java.md:
```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+
+* Bad Security
+
+Remember at the beginning of the section, a part mentioned about accessing the private fields of a class using
```
Maybe can change to reference your applications section since the example given there is more concrete?

leeyimin reviewed (changes requested):
Some minor changes suggested! Is it possible to not change your commit history? It would make it easier to review.

MightyCupcakes committed:
Implement requested changes

leeyimin commented on contents/java/Java.md:
````diff
@@ -99,7 +100,7 @@ An example implementation (not the actual) of the `Subscribe` annotation might b
 ```
 
-And that is it! Important parts of the code to note is the first and second line before the declaration. The first line
+And that is it! Important parts of the code to note are the first and second line before the declaration. The first line
````
first and second lines seems more correct?

MightyCupcakes committed:
Change sentence structure of sentence pointed out by reviewer

leeyimin reviewed (approved):
LGTM!

damithc commented on contents/java/Java.md:
```diff
@@ -14,6 +14,287 @@ Authors: John Doe, Jane Doe
 ## Reflection
 
-// Reflection overview
+### What is Java reflections
```
May be put this topic in a separate document altogether? It's unlikely we can put all Java stuff in one file anyway.

damithc commented on contents/java/Java.md:
```diff
@@ -14,6 +14,287 @@ Authors: John Doe, Jane Doe
+
+Java reflections allow the inspection of classes, methods and fields during runtime, without having any knowledge of it
+during compile time. It is also possible to create new instance of existing objects, invoke methods and change values 
+of fields of existing objects. Hence the name reflection because Java is in a way looking at itself during runtime!
```
Give a motivating example? e.g. to access a private method during testing. As in presentations, an early WIIFY can motivate the reader to continue reading.

damithc commented on contents/java/Java.md:
```diff
@@ -14,6 +14,287 @@ Authors: John Doe, Jane Doe
+
+### The Basics of Reflections
+
+Before getting started with reflections in Java, it is important to realise that classes themselves is an object. Every
```
`realise` -> for consistency, we use American spelling

damithc commented on contents/java/Java.md:
````diff
@@ -14,6 +14,287 @@ Authors: John Doe, Jane Doe
+		return gender;
+	}
+}
+```
````
The above code example is longer than necessary?

damithc commented on contents/java/Java.md:
```diff
@@ -14,6 +14,287 @@ Authors: John Doe, Jane Doe
+.The event bus uses reflections to observe all registered objects via `register` method for methods annotated with the
+`Subscribe` annotation.
+
+An example implementation (not the actual) of the `Subscribe` annotation might be:
```
I'm not sure if annotations is the example you want to start with. Annotations itself is a separate topic and the way it uses reflections is not even visible to the user.

damithc commented on contents/java/Java.md:
```diff
@@ -14,6 +14,287 @@ Authors: John Doe, Jane Doe
+For example you might have come across this problem where your program crashed and you get a `NullPointerException`
+error in your crash log. As you might have experienced already, runtime errors are more troublesome in that they
+are harder to catch and debug. They might even bring your whole software under the water with it by crashing
+the whole thing.
```
Is this paragraph is meant to be a part of the bullet point, do this instead.
```
* bullet point<br>
paragraph text...
```

damithc commented on contents/java/Java.md:
````diff
@@ -14,6 +14,287 @@ Authors: John Doe, Jane Doe
+    }
+}
+```
+
````
Where possible, shorten code examples by omitting part of the code that is not relevant to the context.

damithc reviewed (changes requested):
Some changes suggested based on a quick read. Not a full review.

MightyCupcakes committed:
Split java reflections to a seperate file and update ToC

MightyCupcakes committed:
Restructure the file according to review

Move event centre example to the bottom and moved the private field
accessor example to the top.

MightyCupcakes committed:
Break big paragraph into two

The huge paragraph gave the impression that it should be seperate points
when it is really not is concerning. Therefore, break this paragraph
into smaller parts.

MightyCupcakes committed:
Remove more code that is not relevant to the example

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,291 @@
+during compile time. It is also possible to create new instance of existing objects, invoke methods and change values 
+of fields of existing objects. Hence the name reflection because Java is in a way looking at itself during runtime!
+
+With reflections, it is possible to access private fields and/or methods of a class without having to modify the class
```
`For example, you can use Java reflection to access a ...`

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,291 @@
+of fields of existing objects. Hence the name reflection because Java is in a way looking at itself during runtime!
+
+With reflections, it is possible to access private fields and/or methods of a class without having to modify the class
+itself. This is very useful if you are interested to implement tests cases to test the class.
```
`... tests cases to test a private method.`

Neurrone commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,291 @@
+
+A good example of reflections is to get a private field of another class. While this should
+optimally be solved by modifying the field visibility to `protected` or `public`, sometimes it is not possible to do so
+becuase you might not have any access to the code (for example codes in libraries or frameworks).
```
becuase -> because

Neurrone commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,291 @@
+optimally be solved by modifying the field visibility to `protected` or `public`, sometimes it is not possible to do so
+becuase you might not have any access to the code (for example codes in libraries or frameworks).
+
+For the sake of simplicity, let us use the previous example of a `Animal`. The class definition can look like this.
```
"let us use the previous example of a  Animal". Was the definition of the animal class moved down?

Neurrone commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,291 @@
+
+        try {
+            Field f = animalClazz.getDeclaredField("age");
+            f.setAccessible(true);
```
Does this field remain accessible after the method finishes executing?

Neurrone commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,291 @@
+
+And that is it! Important parts of the code to note are the first two lines before the declaration. The first line
+tells Java that this annotation must not be discarded during compile time so it will be available during runtime. The
+retention poilcy is there because some annotation do not mean anything after compilation (such as `Override` and
```
poilcy -> policy

Neurrone commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,291 @@
+
+Java reflections allow the inspection of classes, methods and fields during runtime, without having any knowledge of it
+during compile time. It is also possible to create new instance of existing objects, invoke methods and change values 
+of fields of existing objects. Hence the name reflection because Java is in a way looking at itself during runtime!
```
Suggest explaining what reflection is without referring to Java first, since it is a general feature that other languages support as well.

MightyCupcakes committed:
Implement suggested changes

damithc commented on contents/ToC.md:
```diff
@@ -3,6 +3,7 @@
 * C#
 * C++
 * [Java](java/Java.md)
+    * [Java Reflections](java/JavaReflections.md)
```
Reflections or Reflection?

damithc commented on contents/java/Java.md:
```diff
@@ -16,4 +16,4 @@ Authors: John Doe, Jane Doe
 // Reflection overview
 
-// Learning resources
+// Learning resources
```
Unintended?

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,298 @@
+way looking at itself during runtime!
+
+With reflections, it is possible to access private fields and/or methods of a class without having to modify the class
+itself. This is very useful if you are interested to implement tests cases to test the class.
```
`test cases`? Also, give this as an example of reflection. The current phrasing sound like it is the primary/sole use of reflection.

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,298 @@
+
+### The Basics of Reflections
+
+Before getting started with reflections in Java, it is important to realize that classes themselves is an object. Every
```
Feels like singular/plural mix-up: `classes themselves is an object`.

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,298 @@
+being made readily available on the web. One good place to start is [http://www.javaworld.com/article/2077015/java-se/take-an-in-depth-look-at-the-java-reflection-api.html](http://www.javaworld.com/article/2077015/java-se/take-an-in-depth-look-at-the-java-reflection-api.html)
+
+What this page will do instead is to give some examples to demonstrate what reflections can do and give you some
+ideas on how you can use reflections in your project.
```
This approach (i.e. not duplicate, point to external resource etc.) is common to all pages in this repo, so no sure we want to explicitly mention it here.

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,298 @@
+What this page will do instead is to give some examples to demonstrate what reflections can do and give you some
+ideas on how you can use reflections in your project.
+
+One important point is that while Java reflections are powerful, its implementations are anything but. There are however
```
Are you saying implementations are not powerful? Or did you intend to say it is not simple?

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,298 @@
+
+Authors: [Jeremy Goh](https://github.com/MightyCupcakes)
+
+## Reflection
```
This page can be entirely about Reflection. i.e. this can be a level 1 heading.

damithc replied:
My bad. The current heading level is fine.

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,298 @@
+setting that particular local `Field` instance to be accessible, not the actual `age` field itself. You can read more
+about it in this [StackOverflow question](http://stackoverflow.com/questions/10638826/java-reflection-impact-of-setaccessibletrue).
+
+#### A More advanced application
```
`more`?

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,298 @@
+
+#### A More advanced application
+
+You might have learnt from your Software Engineering module that the observer pattern can be used for objects that are
```
`Observer`?

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,298 @@
+Compile time errors are easy to catch. Whenever you compile your code, the compiler cleverly spots any error you
+missed and points it out (along with line number and other useful information) to you before quitting. But by using
+reflections, you are bypassing these checks because there is no way to check such errors during compile time. These 
+uncaught errors may cause your program to fail during runtime instead, turning into runtime errors.
```
Indent this paragraph by 2 spaces if you want it to appear as part of the previous bullet point.

damithc commented on contents/java/JavaReflections.md:
```diff
@@ -0,0 +1,298 @@
+* A short but precise overview of Java reflections
+
+[http://www.oracle.com/technetwork/articles/java/javareflection-1536171.html](http://www.oracle.com/technetwork/articles/java/javareflection-1536171.html)
+
```
Here too, you can indent to make the link appear as part of the bullet point.

damithc reviewed (commented):
Good work @MightyCupcakes 
Some comments added. Revise as you see fit.
Adding more content and/or getting more peer reviews is welcome but not required. 
I will be merging end of week 12

MightyCupcakes committed:
Implement requested changes

damithc committed:
Merge branch 'master' into java-reflections
